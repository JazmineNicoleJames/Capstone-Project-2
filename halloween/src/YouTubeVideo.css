.result-text {
    font-family: 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif;
}

h3.result-text {
    font-size: smaller;
}


.title {
    font-size: 15px;
    display: flexbox;
    flex-direction: column;
}