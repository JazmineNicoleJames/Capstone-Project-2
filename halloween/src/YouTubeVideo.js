import React, {useState, useEffect } from "react";
import { useParams } from "react-router-dom";
import HalloweenApi from "./api";
import Questions from "./Questions";
import "./YouTubeVideo.css";


/* YouTubeVideo component
  * holds state for --
    * videoData (data from Youtube API)
    * title (data from the Youtube API -- title of youtube video)
    * description (data from Youtube API -- video description, possibly containing products)
  
  * calls on Halloween API (getCostume) to fetch data about specific costume (item_name)

  * renders YouTube video

*/

const YouTubeVideo = () => {
  const { item_name } = useParams();
  const [videoData, setVideoData] = useState(null);
  const [title, SetTitle] = useState(null);
  const [description, setDescription] = useState(null);
  const [error, setError] = useState(null);

  useEffect(() => {
    async function fetchData() {
      try {
        if(item_name){
          const res = await HalloweenApi.getCostume(item_name);
          if(res && res.snippet && res.snippet.description !== undefined){
            let description = res.snippet.description;
            let videoTitle = res.snippet.title;
            SetTitle(videoTitle)
            const videoId = res.id;
            let videoData = `https://www.youtube.com/embed/${videoId}`;
            setVideoData(videoData)
            if(description.indexOf("Products") !== null){
              let descriptionIdx = description.indexOf("Products");
              setDescription(description.substring(descriptionIdx))
            }
          } else{
            return(<Questions />)
          }
        }


      } catch (e) {
        setError(e)
/*         console.error('Error fetching video data', e) */
      }
    }
    fetchData();
  }, [item_name]);

  if(error){
    return( <div> Error </div>)
  }

    return (
        <div>
        <h3 className="result-text"> Watch this tutorial to learn how to transform yourself into </h3>
        {item_name && (
        <p className="result-text">{item_name.replace(/-/g, ' ').toUpperCase()}</p>
        )}
        <iframe
        width="560"
        height="315"
        src={videoData}
        frameborder="0"
        suppressHydrationWarning={true}
        allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
      ></iframe>
              <p className="title">{title}</p>
              {description && (
                <div style={{ maxHeight:"200px", overflowY:"auto"}}>
                  <ul>
                  {description.split("\n").map((item, index) => (
                    <li key={index}>{item}</li>
                  ))} 
                    </ul></div>
              )}
      </div>

    )
};

export default YouTubeVideo;