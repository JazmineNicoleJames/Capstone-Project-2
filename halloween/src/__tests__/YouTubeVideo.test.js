import React from "react";
import { render, waitFor } from '@testing-library/react';
import { MemoryRouter, Routes, Route } from "react-router-dom";
import YouTubeVideo from "../YouTubeVideo";
import HalloweenApi from "../api";

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useParams: () => ({
        item_name: 'fake'
    })
}));

jest.mock('../api', () => ({
    __esModule: true,
    default: {
        getCostume: jest.fn().mockResolvedValue(
            {id: 'fake',
                snippet:{
                    title: 'mock video title',
                    description: 'fake'
                }
            })
    }
}));


test("should match snapshot", async () => {
    const costume = {
        id: 'fake',
            snippet:{
                title: 'mock video title',
                description: 'fake'
            }
    }
    const { asFragment } = render(
        <MemoryRouter initialEntries={['/youtube-video/fake']}>
            <Routes>
                <Route path='/youtube-video/:item_name' element={<YouTubeVideo allCostumes={costume}/>} />
            </Routes>
        </MemoryRouter>
    );
    
    await waitFor(() => {
        expect(HalloweenApi.getCostume).toHaveBeenCalledWith('fake');
        expect(asFragment()).toMatchSnapshot();
    });
});

test("render youTube video", async () => {
    const { getByText } = render(
        <MemoryRouter initialEntries={['/youtube-video/fake']}>
            <Routes>
                <Route path='/youtube-video/:item_name' element={<YouTubeVideo/>} />
            </Routes>
        </MemoryRouter>
    );
    
    await waitFor(() => {  
        expect(getByText("Watch this tutorial to learn how to transform yourself into")).toBeInTheDocument()
    });
});
