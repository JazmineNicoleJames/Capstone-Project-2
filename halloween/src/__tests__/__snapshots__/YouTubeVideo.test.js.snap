// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <h3
      class="result-text"
    >
       Watch this tutorial to learn how to transform yourself into 
    </h3>
    <p
      class="result-text"
    >
      FAKE
    </p>
    <iframe
      allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
      frameborder="0"
      height="315"
      width="560"
    />
    <p
      class="title"
    />
  </div>
</DocumentFragment>
`;
